hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/WritableKVStateBase.java

```diff
@@ -0,0 +1,265 @@
+import java.util.*;
+
+/**
+ * A base class for implementations of {@link WritableKVState}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public abstract class WritableKVStateBase<K extends Comparable<K>, V>
+        extends ReadableKVStateBase<K, V> implements WritableKVState<K, V> {
+    /** A map of all modified values buffered in this mutable state */
+    private final Map<K, V> modifications = new HashMap<>();
+
+    /**
+     * Create a new StateBase.
+     *
+     * @param stateKey The state key. Cannot be null.
+     */
+    protected WritableKVStateBase(@NonNull final String stateKey) {
+        super(stateKey);
+    }
+
+    /**
+     * Flushes all changes into the underlying data store. This method should <strong>ONLY</strong>
+     * be called by the code that created the {@link WritableKVStateBase} instance or owns it. Don't
+     * cast and commit unless you own the instance!
+     */
+    public final void commit() {
+        for (final var entry : modifications.entrySet()) {
+            final var key = entry.getKey();
+            final var value = entry.getValue();
+            if (value == null) {
+                removeFromDataSource(key);
+            } else {
+                putIntoDataSource(key, value);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>Clears the set of modified keys and removed keys. Equivalent semantically to a "rollback"
+     * operation.
+     */
+    @Override
+    public final void reset() {
+        super.reset();
+        modifications.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nullable
+    public final V get(@NonNull K key) {
+        // If there is a modification, then we've already done a "put" or "remove"
+        // and should return based on the modification
+        if (modifications.containsKey(key)) {
+            final var value = modifications.get(key);
+            super.markRead(key, value);
+            return value;
+        } else {
+            return super.get(key);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nullable
+    public final V getForModify(@NonNull final K key) {
+        Objects.requireNonNull(key);
+        // If there is a modification, then we've already done a "put" or "remove"
+        // and should return based on the modification
+        if (modifications.containsKey(key)) {
+            return modifications.get(key);
+        }
+
+        // If the modifications map does not contain an answer, but the read cache of the
+        // super class does, then it means we've looked this up before but never modified it.
+        // So we can just delegate to the super class.
+        if (hasBeenRead(key)) {
+            return super.get(key);
+        }
+
+        // We have not queried this key before, so let's look it up and store that we have
+        // read this key. And then return the value.
+        final var val = getForModifyFromDataSource(key);
```

Comment:
I think we do. It is discussed in the video. In short, getForModify tells the underlying data source that there is a high degree of likelihood that this get will be followed by a put for the same key. This allows us to do some optimizations on the backend. I wish we didn't need it, because it exposes an implementation detail in the API (which is generally a bad thing to do).